### PR TITLE
Make routing_api component optional

### PR DESCRIPTION
**Pull request by qu1queee**
This PR adds a new feature flag for making the `routing_api` component optional for the CF deployments.

## Description
The rational behind, is that the `routing_api` exposes an API to register routes, which is mainly needed for external clients. Internal clients use `nats`. Therefore, for the CF deployments where none external client are present, disabling the `routing_api` makes sense.

Also, the `routing_api`  generate SSE events that are used by the `tcp-router` instance_group, when the `routing_api` is disabled, we also need to disable the `tcp-router` instance_group.

## Motivation and Context
Same as the above **Description**

## How Has This Been Tested?
I tested this by deploying it with an external DB. The deployment with the local DB never worked for me(seems the seeding have problems).

Running with diego, smoke passes, but not CAT´s. For CAT´s everything looks fine, the only details is that the `Tcp Routing` tests failed. I´m not sure how to deal in this PR with tests that are required to be avoided.

## Screenshots (if appropriate):
Screenshot 2020-03-24 at 16 54 42

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code has security implications.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml [hunk 1]
```diff
@@ -28,3 +31,35 @@
...
+- type: remove
+  path: /addons/name=bosh-dns-aliases/jobs/name=bosh-dns-aliases/properties/aliases/domain=routing-api.service.cf.internal
+
+- type: remove
+  path: /variables/name=routing_api_tls
+
+- type: remove
+  path: /variables/name=routing_api_tls_client
+
+# Disable the tcp-router ig, while there will be none routing-api SSE events to read from.
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml]:
```suggestion
# Disable the tcp-router ig, as there will be no routing-api SSE events to read from.
```

Why is this here, and we _also_ have things in `tcp-router.yaml` that depend on the same flag?

File: deploy/helm/kubecf/assets/operations/instance_groups/tcp-router.yaml [hunk 2]
```diff
@@ -1,9 +1,12 @@
 {{- $service := index .Values.services "tcp-router" }}
 
+{{- if .Values.features.routing_api.enabled }}
 - type: replace
   path: /instance_groups/name=routing-api/jobs/name=routing-api/properties/routing_api/router_groups/name=default-tcp/reservable_ports
   value: "{{ $service.port_range.start }}-{{ $service.port_range.end }}"
+{{- end }}
 
+{{- if .Values.features.routing_api.enabled }}
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/tcp-router.yaml]:
```suggestion

```

Why are we terminating the condition and then immediately restarting it?

**Comment by qu1queee** [comment]
@mook-as thanks for the review, I will refactor the PR soon. I would be interested to hear @viovanov opinion on this

**Comment by viovanov** [comment]
@qu1queee there are settings in CATs to disable test suites.
You should tweak those so CATs pass with the routing api off.

I think this is a fair use case, however I'd like to see a PR opened on https://github.com/cloudfoundry/cf-deployment/ to add this operation to upstream.

I wouldn't block this if this is urgent for you @qu1queee, but if it's not, we should do the right thing and contribute it to upstream.

**Comment by qu1queee** [comment]
@mook-as @viovanov I addressed your comments and PR requested changes. It would be helpful to have this in KubeCF for now.

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml [hunk 3]
```diff
@@ -28,3 +31,35 @@
...
+
+- type: remove
+  path: /variables/name=routing_api_tls
+
+- type: remove
+  path: /variables/name=routing_api_tls_client
+
+# Disable the tcp-router ig, as there will be no routing-api SSE events to read from.
+- type: remove
+  path: /instance_groups/name=tcp-router
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml]: Shouldn't this live in `tcp-router.yaml`?

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml [hunk 4]
```diff
@@ -28,3 +31,35 @@
...
 {{ $root.Files.Get $path }}
 {{- end }}
+
+{{- else }}
+# Disable routing-api.
+- type: remove
+  path: /instance_groups/name=api/jobs/name=routing-api
+
+- type: remove
+  path: /instance_groups/name=database?/jobs/name=pxc-mysql/properties/seeded_databases/name=routing-api
```
[the remark this answers is not in the document]
Comment (qu1queee) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:41]: @mook-as you can do that, but independently of the DB type in KubeCF, you need to remove this, while it is a property that lives directly in https://github.com/cloudfoundry/cf-deployment/blob/v12.33.0/cf-deployment.yml#L375-L377

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml (hunk 3 above)
Comment (qu1queee) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:58]: yes indeed, @mook-as pls request changes, Im rerequesting you a review.!

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mook-as) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:41]: Ah! I got confused by the `name=database?`; looking at [`database.yaml`](https://github.com/cloudfoundry-incubator/kubecf/blob/master/deploy/helm/kubecf/assets/operations/instance_groups/database.yaml#L2) we actually unconditionally remove the instance group. We do have a separate db-seeder that you've already covered in this PR.

That is, this is just a long-winded way to ask: why do we need the `?` on that path? Is it becase that instance group has _already_ been removed, and there's no point in removing it again?

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml (hunk 3 above)
Comment (mook-as) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:58]: Request changes it is then!

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (qu1queee) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:41]: yes

[the remark this answers is not in the document]
Comment (mook-as) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:41]: In that case, shouldn't we drop the whole operation (since it's, essentially, removing something that doesn't exist)?

[the remark this answers is not in the document]
Comment (qu1queee) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:41]: @mook-as sorry, I misunderstood your question. The `?` was just in case the DB was removed, we removed the DB in KubeCF in https://github.com/cloudfoundry-incubator/kubecf/blob/master/deploy/helm/kubecf/assets/operations/instance_groups/database.yaml#L2-L3 , but the `?` needs to be there as a security measure, unless you know the interpolation order when applying the different YAMLS from https://github.com/cloudfoundry-incubator/kubecf/tree/master/deploy/helm/kubecf/assets/operations/instance_groups on top of the cf-deployment manifest. In other words, how to know what happens first, https://github.com/cloudfoundry-incubator/kubecf/blob/master/deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml vs https://github.com/cloudfoundry-incubator/kubecf/blob/master/deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml ?

**Comment by qu1queee** [comment]
@mook-as also, this is the output when I ran the above `routing_api` ops-file, without the `?`under:
```yaml
- type: remove
  path: /instance_groups/name=database/jobs/name=pxc-mysql/properties/seeded_databases/name=routing-api
```

output:
```yaml
Running: helm --kubeconfig /var/folders/6p/66zfm2sn5xv8x_q9yz7q2h8c0000gn/T/bkcf-cluster-03909124851/kubeConfig884692699/kube-config-dal10-bkcf-cluster-03.yml install kubecf .../kubecf/bazel-bin/deploy/helm/kubecf/kubecf.tgz --namespace kubecf --wait -f /var/folders/6p/66zfm2sn5xv8x_q9yz7q2h8c0000gn/T/value-overrides551616925 -f /var/folders/6p/66zfm2sn5xv8x_q9yz7q2h8c0000gn/T/value-overrides441484888 -f /var/folders/6p/66zfm2sn5xv8x_q9yz7q2h8c0000gn/T/value-overrides706725335
╭ Error: failed to install kubecf
│ Error: admission webhook "validate-boshdeployment.quarks.cloudfoundry.org" denied
│ the request: Failed to resolve manifest: Failed to interpolate ops
│ 'kubecf-ops-routing-api' for manifest 'kubecf': Applying ops on manifest obj
│ failed in interpolator: Expected to find exactly one matching array item for path
│ '/instance_groups/name=database' but found 0
│ : exit status 1
╵

```

**Comment by qu1queee** [comment]
@mook-as pls see my comments, its getting hard to keep this PR updated with so many changes in master branch, can you pls prioritize this one?

File: deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mook-as) [reply deploy/helm/kubecf/assets/operations/instance_groups/routing-api.yaml:41]: Ah! I thought we sorted the glob, but we don't, so we'll have to do the guard. Thanks!
